app/src/main/java/org/thoughtcrime/securesms/util/AlarmSleepTimer.java

```diff
@@ -0,0 +1,98 @@
+package org.thoughtcrime.securesms.util;
+
+import android.app.AlarmManager;
+import android.app.PendingIntent;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.Build;
+import android.os.SystemClock;
+
+import org.signal.core.util.logging.Log;
+import org.whispersystems.signalservice.api.util.SleepTimer;
+
+import java.util.concurrent.ConcurrentSkipListSet;
+
+/**
+ * A sleep timer that is based on elapsed realtime, so
+ * that it works properly, even in low-power sleep modes.
+ *
+ */
+public class AlarmSleepTimer implements SleepTimer {
+  private static final String TAG = Log.tag(AlarmSleepTimer.class);
+  private static ConcurrentSkipListSet<Integer> actionIdList = new ConcurrentSkipListSet<>();
+
+  private final Context context;
+
+  public AlarmSleepTimer(Context context) {
+    this.context = context;
+  }
+
+  @Override
+  public void sleep(long millis) {
+    final AlarmReceiver alarmReceiver = new AlarmSleepTimer.AlarmReceiver();
+    int actionId = 0;
+    while (!actionIdList.add(actionId)){
+      actionId++;
+    }
+    try {
+      context.registerReceiver(alarmReceiver,
+             new IntentFilter(AlarmReceiver.WAKE_UP_THREAD_ACTION + "." + actionId));
+
+      final long startTime = System.currentTimeMillis();
+      alarmReceiver.setAlarm(millis, AlarmReceiver.WAKE_UP_THREAD_ACTION + "." + actionId);
+
+      while (System.currentTimeMillis() - startTime < millis) {
+        try {
+          synchronized (this) {
+            wait(millis - System.currentTimeMillis() + startTime);
+          }
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)